<launch>
    <!-- RS args -->
    <arg name="serial_no"             default=""/>
    <arg name="json_file_path"        default=""/>
    <arg name="camera"                default="camera"/>
    <arg name="manager"             default="realsense2_camera_manager"/>

    <arg name="depth_width"       value="640"/>
    <arg name="depth_height"      value="480"/>
    <arg name="depth_fps"         value="30"/>
    <arg name="color_width"       value="640"/>
    <arg name="color_height"      value="480"/>
    <arg name="color_fps"         value="30"/>
    <arg name="enable_depth"      value="true"/>
    <arg name="enable_color"      value="true"/>
    <arg name="enable_pointcloud" value="false"/>
    <arg name="align_depth"       value="true"/>
    <arg name="enable_sync"       value="true"/>
    <arg name="tf_prefix"         value="$(arg camera)"/>

    <!-- Arguments for remapping all device namespaces -->
    <arg name="rgb"                             default="color" />
    <arg name="ir"                              default="infra1" />
    <arg name="depth"                           default="depth" />
    <arg name="depth_registered_pub"            default="depth_registered" />
    <arg name="depth_registered"                default="depth_registered" unless="$(arg align_depth)" />
    <arg name="depth_registered"                default="aligned_depth_to_color" if="$(arg align_depth)" />
    <arg name="depth_registered_filtered"       default="$(arg depth_registered)" />
    <arg name="projector"                       default="projector" />

    <!-- Disable bond topics by default -->
    <arg name="bond"                            default="false" />
    <arg name="respawn"                         default="$(arg bond)" />

    <!-- Processing Modules -->
    <arg name="rgb_processing"                  default="true"/>
    <arg name="debayer_processing"              default="false" />
    <arg name="ir_processing"                   default="false"/>
    <arg name="depth_processing"                default="false"/>
    <arg name="depth_registered_processing"     default="true"/>
    <arg name="disparity_processing"            default="false"/>
    <arg name="disparity_registered_processing" default="false"/>
    <arg name="hw_registered_processing"        default="$(arg align_depth)" />
    <arg name="sw_registered_processing"        default="true" unless="$(arg align_depth)" />
    <arg name="sw_registered_processing"        default="false" if="$(arg align_depth)" />



    <group ns="$(arg camera)">
    <include file="$(find realsense2_camera)/launch/includes/nodelet.launch.xml">
        <arg name="serial_no"         value="$(arg serial_no)"/>
        <arg name="json_file_path"    value="$(arg json_file_path)"/>
        <arg name="depth_width"       value="640"/>
        <arg name="depth_height"      value="480"/>
        <arg name="depth_fps"         value="30"/>
        <arg name="color_width"       value="640"/>
        <arg name="color_height"      value="480"/>
        <arg name="color_fps"         value="30"/>
        <arg name="enable_depth"      value="true"/>
        <arg name="enable_color"      value="true"/>
        <arg name="enable_pointcloud" value="false"/>
        <arg name="align_depth"       value="true"/>
        <arg name="enable_sync"       value="true"/>
    <arg name="tf_prefix"         value="$(arg camera)"/>
    </include>
    
    <!-- RGB processing -->
    <include if="$(arg rgb_processing)" file="$(find rgbd_launch)/launch/includes/rgb.launch.xml">
        <arg name="manager"                       value="$(arg manager)" />
        <arg name="respawn"                       value="$(arg respawn)" />
        <arg name="rgb"                           value="$(arg rgb)" />
        <arg name="debayer_processing"            value="$(arg debayer_processing)" />
    </include>

    <group if="$(eval depth_registered_processing and sw_registered_processing)">
        <node pkg="nodelet" type="nodelet" name="register_depth" args="load depth_image_proc/register $(arg manager) $(arg bond)" respawn="$(arg respawn)">
        <remap from="rgb/camera_info"             to="$(arg rgb)/camera_info" />
        <remap from="depth/camera_info"           to="$(arg depth)/camera_info" />
        <remap from="depth/image_rect"            to="$(arg depth)/image_rect_raw" />
        <remap from="depth_registered/image_rect" to="$(arg depth_registered)/sw_registered/image_rect_raw" />
        </node>

        <!-- Publish registered XYZRGB point cloud with software registered input -->
        <node pkg="nodelet" type="nodelet" name="points_xyzrgb_sw_registered" args="load depth_image_proc/point_cloud_xyzrgb $(arg manager) $(arg bond)" respawn="$(arg respawn)">
        <remap from="rgb/image_rect_color"        to="$(arg rgb)/image_rect_color" />
        <remap from="rgb/camera_info"             to="$(arg rgb)/camera_info" />
        <remap from="depth_registered/image_rect" to="$(arg depth_registered_filtered)/sw_registered/image_rect_raw" />
        <remap from="depth_registered/points"     to="$(arg depth_registered)/points" />
        </node>
    </group>

    <group if="$(eval depth_registered_processing and hw_registered_processing)">
        <!-- Publish registered XYZRGB point cloud with hardware registered input (ROS Realsense depth alignment) -->
        <node pkg="nodelet" type="nodelet" name="points_xyzrgb_hw_registered"
            args="load depth_image_proc/point_cloud_xyzrgb $(arg manager) $(arg bond)" respawn="$(arg respawn)">
        <remap from="rgb/image_rect_color"        to="$(arg rgb)/image_rect_color" />
        <remap from="rgb/camera_info"             to="$(arg rgb)/camera_info" />
        <remap from="depth_registered/image_rect" to="$(arg depth_registered)/image_raw" />
        <remap from="depth_registered/points"     to="$(arg depth_registered_pub)/points" />
        </node>
    </group>
    </group>

</launch>
